be/src/olap/delta_writer.cpp

```diff
@@ -132,48 +135,93 @@ OLAPStatus DeltaWriter::init() {
 
     _tablet_schema = &(_tablet->tablet_schema());
     _schema = new Schema(*_tablet_schema);
-    _mem_table = new MemTable(_schema, _tablet_schema, _req.slots,
-                              _req.tuple_desc, _tablet->keys_type());
+    _mem_table = std::make_shared<MemTable>(_tablet->tablet_id(), _schema, _tablet_schema, _req.slots,
+            _req.tuple_desc, _tablet->keys_type(), rowset_writer());
+
+    // create flush hanlder for flushing memtable
+    RETURN_NOT_OK(_storage_engine->memtable_flush_executor()->create_handler(
+            _tablet->data_dir()->path_hash(), &_flush_handler));
+
     _is_init = true;
     return OLAP_SUCCESS;
 }
 
 OLAPStatus DeltaWriter::write(Tuple* tuple) {
     if (!_is_init) {
-        auto st = init();
-        if (st != OLAP_SUCCESS) {
-            return st;
-        }
+        RETURN_NOT_OK(init());
     }
 
     _mem_table->insert(tuple);
+
+    // if memtable is full, push it to the flush executor,
+    // and create a new memtable for incoming data
     if (_mem_table->memory_usage() >= config::write_buffer_size) {
-        RETURN_NOT_OK(_mem_table->flush(_rowset_writer.get()));
+        RETURN_NOT_OK(_flush_memtable_async());
+        // create a new memtable for new incoming data
+        _mem_table.reset(new MemTable(_tablet->tablet_id(), _schema, _tablet_schema, _req.slots,
+                _req.tuple_desc, _tablet->keys_type(), rowset_writer()));
+    }
+    return OLAP_SUCCESS;
+}
+
+OLAPStatus DeltaWriter::_flush_memtable_async() {
+    RETURN_NOT_OK(_check_flush_futures(false));
+    _flush_future_queue.push(std::move(_flush_handler->process(_mem_table)));
+    return OLAP_SUCCESS;
+}
 
-        SAFE_DELETE(_mem_table);
-        _mem_table = new MemTable(_schema, _tablet_schema, _req.slots,
-                                  _req.tuple_desc, _tablet->keys_type());
+OLAPStatus DeltaWriter::_check_flush_futures(bool block) {
```

Comment:
why not put this logic in _flush_handler. DeltaWriter can test status of _flush_handler. 
Executor can update flush_handler's status through callback. In that way, there is no need to loop the status of submitted MemTable. 

In your implementation code, you give a 1 microseconds interval to loop the status. When waiting last MemTable, block is true, you will loop 10^6 times one second, which is waste of CPU.

I think following interface is enough for current usage.
```
class FlusherClient {
public:
    // submit a memtable to flush. return error if some previous submitted MemTable has failed  
    Status submit();
    // wait for all submitted memtable finished.
    Status wait();
    // Get flush operations' statistics
    Statistics get_stats();

    // called when a memtable is finished by executor.
    void on_flush_finished();
};
```